os/board/rtl8730e/src/component/os_dep/osdep_service_queue.c

```diff
@@ -29,33 +32,139 @@ u32 rtw_end_of_queue_search(_list *head, _list *plist)
 }
 int rtw_init_xqueue(_xqueue *queue, const char *name, u32 message_size, u32 number_of_messages)
 {
-	/* To avoid gcc warnings */
-	DBG_INFO("\n");
-	return 0;
+	mqd_t pmqd;
+	/* The following character array of size 9 is used to store pointer address in hex represented as ascii.
+	   As address has 32-bit integer value, the maximum string length in hex is 8 with 1 null character.*/
+	char mq_name[9];
+	struct mq_attr attr;
+	int ret;
+	if (!queue) {
+		dbg("queue is NULL\n");
+		return FAIL;
+	}
+	if (*queue) {
+		dbg("msg queue already init\n");
+		return FAIL;
+	}
+
+	attr.mq_maxmsg = number_of_messages * 5;
+	attr.mq_msgsize = message_size > MQ_MAX_BYTES? MQ_MAX_BYTES : message_size;
+
+	itoa((int) queue, mq_name, 16);
+	pmqd = mq_open((const char *) mq_name, O_RDWR | O_CREAT, 0666, &attr);
+	if (pmqd == (mqd_t) ERROR) {
+		ret = get_errno();
+		dbg("mq open fail: %d\n", ret);
+		return FAIL;
+	}
+
+	*queue = pmqd;
+	return SUCCESS;
 }
 
 int rtw_push_to_xqueue(_xqueue *queue, void *message, u32 timeout_ms)
 {
-	DBG_INFO("\n");
-	return 0;
+	int prio = MQ_PRIO_MAX;
+	int ret;
+	if (!queue) {
+		dbg("queue is NULL\n");
+		return FAIL;
+	}
+
+	if (timeout_ms != 0xFFFFFFFF) {
+		struct timespec ts;
+		clock_gettime(CLOCK_REALTIME, &ts);
+		ts.tv_sec += timeout_ms / 1000;
+		ts.tv_nsec += (timeout_ms % 1000) * 1000 * 1000;
+		if (ts.tv_nsec >= NSEC_PER_SEC) {
+			ts.tv_sec ++;
+			ts.tv_nsec -= NSEC_PER_SEC;
+		}
+		if (mq_timedsend((mqd_t) *queue, message, ((mqd_t) *queue)->msgq->maxmsgsize, prio, &ts) != OK) {
+			ret = get_errno();
+			dbg("mq time send fail: %d\n", ret);
+			return FAIL;
+		}
+	} else {
+		if (mq_send((mqd_t) *queue, message, ((mqd_t) *queue)->msgq->maxmsgsize, prio) != OK) {
+			ret = get_errno();
+			dbg("mq send fail: %d\n", ret);
+			return FAIL;
+		}
+	}
+	return SUCCESS;
 }
 
 int rtw_pop_from_xqueue(_xqueue *queue, void *message, u32 timeout_ms)
 {
-	DBG_INFO("\n");
-	return 0;
+	int prio;
+	int ret;
+	if (!queue) {
+		dbg("queue is NULL\n");
+		return FAIL;
+	}
+
+	if (timeout_ms != 0xFFFFFFFF) {
+		struct timespec ts;
+		clock_gettime(CLOCK_REALTIME, &ts);
+		ts.tv_sec += timeout_ms / 1000;
+		ts.tv_nsec += (timeout_ms % 1000) * 1000 * 1000;
+		if (ts.tv_nsec >= NSEC_PER_SEC) {
+			ts.tv_sec ++;
+			ts.tv_nsec -= NSEC_PER_SEC;
+		}
+		if (mq_timedreceive((mqd_t) *queue, message, ((mqd_t) *queue)->msgq->maxmsgsize, &prio, &ts) == ERROR) {
```

Comment:
Is there any specific reason why use mq among the kernel features? if you added this for yout testing, then please add some define or config only for your test

Reply:
Hi @Taejun-Kwon ,

In USB operation, we used queue in our FreeRTOS OS, an alternative in Nuttx is mq.  Would there be issue using this?

Kai

Reply:
because this will share some of result with Virtual driver of USB(dev/usb or something), I think callback mechanism is enough in kernel side.  Chracter Device, should share result with Application message queue because only IPC allowed between kernel & app

Reply:
Hi Taejun,

I will try a work around.

Kai